python/arc/autolinker/autolinker.py

```diff
@@ -225,51 +243,37 @@ def _generate_candidate_blocking_rules(
     data:pyspark.sql.DataFrame,
     attribute_columns:list,
     comparison_size_limit:int,
-    unique_id:str,
-    max_columns_per_rule:int=2
+    max_columns_per_and_rule:int=2,
+    max_rules_per_or_rule:int=3,
   ) -> list:
     """
     Method to automatically generate a list of lists of blocking rules to test, given a user-defined limit for
-    pair-wise comparison size.
+    pair-wise comparison size. Uses appriximated method built in to ARC's SQL functions.
 
     Returns a nested list of lists with Splink-compatible blocking rule queries.
 
     :param data: input data with record-per-row
     :param attribute_columns: valid column names of data, containing all possible columns to block on
     :param comparison_size_limit: the maximum number of pairs we want to compare, to limit hardware issues
-    :param unique_id: the name of the unique ID column in data
-    :param max_columns_per_rule: the maximum number of column comparisons in a single rule to try
+    :param max_columns_per_and_rule: the maximum number of column comparisons in a single rule to try
+    :param max_rules_per_or_rule: the maximum number of rules comparisons in a composite rule to try
 
     """
 
-    # initialise empty list to store all combinations in
-    blocking_combinations = []
+    # replace null values with dummy
+    data = data.fillna("null_")
 
-    # create list of all possible rules
-    for n in range(1, max_columns_per_rule+1):
-      combs = list(itertools.combinations(attribute_columns, n))
-      blocking_combinations.extend(combs)
+    # Create blocking rules and size esimates
+    df_rules = arcf.arc_generate_blocking_rules(data, max_columns_per_and_rule, max_rules_per_or_rule, *attribute_columns)
 
-    # Convert them to Splink-compatible rules
-    blocking_rules = self._generate_rules(["&".join(r) for r in blocking_combinations])
-    
-    # Estimate number of pairs per blocking rule by grouping on blocking rule columns and getting the sum of squares of counts
-    comp_size_dict = dict()
+    # Filter on max
+    rules = df_rules.filter(f"rule_squared_count < {comparison_size_limit}").collect()
 
-    for comb, rule in zip(blocking_combinations, blocking_rules):
-      num_pairs = data.groupBy(list(comb)).count().select(F.sum(F.col("count")*F.col("count"))).collect()[0]["sum((count * count))"]
-      comp_size_dict.update({rule: num_pairs})
-    
-    # loop through all combinations of combinations and save those which remain under the limit (this is a bit brute force)
-    accepted_rules = []
-    for r in range(1, len(blocking_rules)+1):
-      for c in itertools.combinations(blocking_rules, r):
-        if sum([v for k, v in comp_size_dict.items() if k in c])<=comparison_size_limit:
-          accepted_rules.append(list(c))
+    accepted_rules = [rule.splink_rule for rule in rules]
 
+    # set deterministic rules to be 500th largest (or largest) blocking rule
```

Comment:
@marcell-ferencz-databricks this comment isnt correct.
this will take randomly 500 then take the largest one from that sample.
if you wanted 500 largest then orderBy should happen before limit(500)

Reply:
my bad, missed the first order by